ddtrace/internal/datastreams/processor.py

```diff
@@ -0,0 +1,270 @@
+                    resp.reason,
+                    resp.read(),
+                    self._agent_endpoint,
+                )
+            else:
+                log.debug("sent %s to %s", _human_size(len(payload)), self._agent_endpoint)
+
+    def periodic(self):
+        # type: () -> None
+
+        with self._lock:
+            serialized_stats = self._serialize_buckets()
+
+        if not serialized_stats:
+            # No stats to report, short-circuit.
+            return
+        raw_payload = {
+            u"Service": self._service,
+            u"TracerVersion": ddtrace.__version__,
+            u"Lang": "python",
+            u"Stats": serialized_stats,
+            u"Hostname": self._hostname,
+        }  # type: Dict[str, Union[List[Dict], str]]
+        if config.env:
+            raw_payload[u"Env"] = six.ensure_text(config.env)
+        if config.version:
+            raw_payload[u"Version"] = six.ensure_text(config.version)
+
+        payload = packb(raw_payload)
+        compressed = gzip.compress(payload, 1)
+        try:
+            self._retry_request(self._flush_stats, compressed)
+        except tenacity.RetryError:
+            log.error("retry limit exceeded submitting pathway stats to the Datadog agent at %s", self._agent_endpoint)
+
+    def shutdown(self, timeout):
+        # type: (Optional[float]) -> None
+        self.periodic()
+        self.stop(timeout)
+
+    def decode_pathway(self, data):
+        # type: (bytes) -> DataStreamsCtx
+        try:
+            hash_value = struct.unpack("<Q", data[:8])[0]
+            data = data[8:]
+            pathway_start_ms, data = decode_var_int_64(data)
+            current_edge_start_ms, data = decode_var_int_64(data)
+            ctx = DataStreamsCtx(self, hash_value, float(pathway_start_ms) / 1e3, float(current_edge_start_ms) / 1e3)
+            # reset context of current thread every time we decode
+            self._current_context.value = ctx
+            return ctx
+        except EOFError:
+            return self.new_pathway()
+
+    def new_pathway(self):
+        # type: () -> DataStreamsCtx
+        now_sec = time.time()
+        ctx = DataStreamsCtx(self, 0, now_sec, now_sec)
+        return ctx
+
+    def set_checkpoint(self, tags):
+        if hasattr(self._current_context, "value"):
+            ctx = self._current_context.value
+        else:
+            ctx = self.new_pathway()
+            self._current_context.value = ctx
+        ctx.set_checkpoint(tags)
+        return ctx
+
+
+class DataStreamsCtx:
+    def __init__(self, processor, hash_value, pathway_start_sec, current_edge_start_sec):
+        # type: (DataStreamsProcessor, int, float, float) -> None
+        self.processor = processor
+        self.pathway_start_sec = pathway_start_sec
+        self.current_edge_start_sec = current_edge_start_sec
+        self.hash = hash_value
+        self.service = six.ensure_text(config._get_service("unnamed-python-service"))
```

Comment:
However, will the service name here match the service name set for the Kafka integration? It seems like it would be confusing to have stats submitted under one name and spans under another? Even if they represent different things about that integration.

Reply:
spoke offline about this, this makes sense.